### PR TITLE
Rebase core master image from debian to distroless

### LINKED ISSUE #70249: Rebase most docker images on gcr.io/distroless/static
EDIT (from @justaugustus - 5/8/20):

### `go-runner` (distroless++) (@dims)

- [x] (https://github.com/kubernetes/kubernetes/pull/90804) New go-runner image for distroless scenarios
- [x] (https://github.com/kubernetes/k8s.io/pull/849) Promote the new `go-runner` image - `v0.1.0`
- [x] (https://github.com/kubernetes/kubernetes/pull/90852) Bump `go-runner` to `v0.1.1`
- [x] (https://github.com/kubernetes/k8s.io/pull/853) Promote `go-runner:v0.1.1` image
- [x] (https://github.com/kubernetes/kubernetes/pull/90674) Switch core master base images from debian to distroless
- [x] (https://github.com/kubernetes/kubernetes/pull/91329) Flexvolume master driver is blocking controller-manager from moving to distroless - https://github.com/kubernetes/kubernetes/issues/78737

### Debian base images (`debian-base`, `debian-iptables`) and `pause` (@justaugustus)

- [x] (https://github.com/kubernetes/kubernetes/issues/58012) Base image (debian-base, debian-iptables) updates & building on K8s Infra
- [x] (https://github.com/kubernetes/kubernetes/issues/90698) 'pause' base image updates & building on K8s Infra
- [x] (https://github.com/kubernetes/kubernetes/pull/88603) Remove `debian-base` image
- [ ] () Remove `debian-iptables` image

---

Forked from https://github.com/kubernetes/kubernetes/issues/40248 (think of this as V2).

`scratch` is not quite sufficient as a base image for static go binaries. Common dependencies include ca-certificates, resolv.conf, hosts, and nsswtich (see https://github.com/kubernetes/kubernetes/issues/69195).  See below for a more complete list.

We need to decide if we want a scratch+config base image (either found or created by us), or a more complete distro (e.g. debian-base, at the cost of more frequent patching). `gcr.io/distroless/static` has also been proposed, or could provide inspiration for a static base.

**Blocked by:** (for the scratch case) https://github.com/kubernetes/kubernetes/issues/61006, for logs redirection.

/kind bug
/cc @ixdy 

---

**Appendix**

Here is a list of `/etc/*` files referenced from the go source (obviously some of these are not applicable):
```
$ grep -rEoIh '/etc/[-a-zA-Z./_]+[a-zA-Z]' | sort -u
/etc/alpine-release
/etc/apache
/etc/apache/mime.types
/etc/certs/ca-certificates.crt
/etc/group
/etc/hosts
/etc/localtime
/etc/mdns.allow
/etc/mime.types
/etc/nsswitch.conf
/etc/openssl/certs
/etc/openssl/certs/ca-certificates.crt
/etc/passwd
/etc/passwd/foo
/etc/passwx
/etc/paths.d/go
/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem
/etc/pki/tls/cacert.pem
/etc/pki/tls/certs
/etc/pki/tls/certs/ca-bundle.crt
/etc/profile
/etc/protocols
/etc/release
/etc/resolv.conf
/etc/security/cacerts
/etc/services
/etc/ssl/ca-bundle.pem
/etc/ssl/cacert.pem
/etc/ssl/cert.pem
/etc/ssl/certs
/etc/ssl/certs/ca-certificates.crt
```

### PR DESCRIPTION
**Pull request by yuwenma**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug

/kind cleanup

> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

This PR is part of the effort on rebasing kubernetes images to distroless/static. See [this KEP](https://github.com/kubernetes/enhancements/pull/900) for details.  

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by yuwenma** [comment]
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-integration

/test pull-kubernetes-kubemark-e2e-gce-big
--

**Comment by justaugustus** [comment]
@kubernetes/sig-testing / @kubernetes/sig-scalability-bugs -- Hey testing + scalability friends, any idea on this one?
We're getting consistent failures for `pull-kubernetes-e2e-gce` + `pull-kubernetes-kubemark-e2e-gce-big` on this PR.

File: build/common.sh [hunk 1]
```diff
@@ -95,9 +95,9 @@ kube::build::get_docker_wrapped_binaries() {
   ### in build/BUILD. And kube::golang::server_image_targets
   local targets=(
     cloud-controller-manager,"k8s.gcr.io/debian-base-${arch}:${debian_base_version}"
-    kube-apiserver,"k8s.gcr.io/debian-base-${arch}:${debian_base_version}"
-    kube-controller-manager,"k8s.gcr.io/debian-base-${arch}:${debian_base_version}"
-    kube-scheduler,"k8s.gcr.io/debian-base-${arch}:${debian_base_version}"
+    kube-apiserver,"gcr.io/distroless/static:latest"
+    kube-controller-manager,"gcr.io/distroless/static:latest"
+    kube-scheduler,"gcr.io/distroless/static:latest"
```
Comment (BenTheElder) [inline build/common.sh]: Will this get the right architecture?

**Comment by BenTheElder** [comment]
Flexvolume tests seem to be failing pretty reliably.

**Comment by BenTheElder** [comment]
KubeMark up is timing out.

Those are both looking suspicious. You should probably dig into the logs / artifact logs further.

File: build/common.sh (hunk 1 above)
Comment (yuwenma) [reply build/common.sh]: I think all the arch will use the same distroless base image. Does distroless image need multi-arch support?

**Comment by yuwenma** [comment]
Thanks @justaugustus !

(First of all, this PR is currently blocked by kubernetes/klog#53, #77904. It can't go in until we fix the duplicated klog items issue)  
 
Here's what I found:

1. This PR touches both the "make release" and " make bazel release" process. I have tested both and they all work well (build the artifact, successfully create a new cluster using the build artifact).
2. For the two failed test, it looks like they only fail due to the change in the "make bazel release" workflow (#77881 #77882). @BenTheElder Any insights what may go wrong?

File: build/common.sh (hunk 1 above)
Comment (BenTheElder) [reply build/common.sh]: I have no idea what it does here to be honest, but previously we explicitly selected an image matching the architecture we are building for.

**Comment by BenTheElder** [comment]
@yuwenma we run the bazel process in CI because we can cache it. I don't see anything obviously wrong with the build but there are a lot of logs and I'm short on time at the moment, hopefully something in the test or node / pod logs shows why the tests failed. They're also SIG Storage (flex volume) / SIG Scalability (kubemar) tests mainly so perhaps asking for help from there.

**Comment by yuwenma** [comment]
> /test pull-kubernetes-integration

Hi Walter, I think this PR will fail prow test without PR #78466 goes in first. #78466 is currently causing the 2k scalability test failures and we are still triaging the root cause.

**Comment by justaugustus** [comment]
@yuwenma -- where are we at with this? Still pending #78466, right?

**Comment by yuwenma** [comment]
> @yuwenma -- where are we at with this? Still pending #78466, right?

Right. Using flag `--log-file` from klog causes performance regressions and the previous fix (kubernetes/klog#69) doesn't resolved the issue (it fixed log duplication but scalability test still fail). 

I think we need to firstly figure out how klog causes the performance regression and then can move these two PRs forward. Previously we thought the root cause is the duplicated logging messages (Error message is logged three times due to the log severity cascading strategy). kubernetes/klog#65 fixed the log duplication but the scalability test still reports performance regression (kubernetes/kubernetes#78734). So we would need some more in-depth research on klog side, but I'm not sure who is available to do that.

The most tricky part is that the performance issue is only caught by the kubernetes scalability test, so the fixing progress is like "A fix attempt to klog" --> "release klog" --> "import to kubernetes" --> "run scalability test( which is not a default prow presubmit tests)". This progress makes the fix takes even longer and it could cost more efforts if the fix doesn't fully work. 

I think the most valuable approaches are 1. to fully understand the klog file logging workflow before any further fixes and 2. add scalability/pressure test on klog instead of totally relying on k/k scalability test to catch the issue.

@dims @justaugustus @liggitt @wojtek-t Thoughts?

**Comment by liggitt** [comment]
Agreed. https://github.com/kubernetes/klog/issues/70 opened to track building out those perf/scale tests in klog.

**Comment by justaugustus** [comment]
@yuwenma @liggitt -- What's the current status on this?
cc: @DirectXMan12 (ref: https://github.com/kubernetes/klog/issues/70)

**Comment by dims** [comment]
/hold

I don't see any evidence that distroless images support architectures other than amd64 (Please correct me if i am wrong!!!)

cc @tallclair

**Comment by BenTheElder** [comment]
oof yeah, inspecting the images in cloud console shows only amd64 manifests (sampled a few).
https://github.com/GoogleContainerTools/distroless/issues/406

**Comment by dims** [comment]
@BenTheElder someone ported to arm arch as well https://github.com/discolix/discolix but that's it other arches don't work either

cc @tallclair

**Comment by dims** [comment]
Looks like i missed a nuance, even if the manifest says amd64 like Ben observed, looks like the key here is `:static` 
https://github.com/GoogleContainerTools/distroless/issues/346

**Comment by BenTheElder** [comment]
We still have to be careful to make sure the release images have the right
architecture metadata.

**Comment by dims** [comment]
/hold cancel

Right @BenTheElder

**Comment by dekkagaijin** [comment]
Seems like `distroless:static` should be trivially multi-arch-friendly, though that obviously doesn't extend to `:base` and others with bundled libs and runtimes. I'll follow up to see how difficult it would be to push a manifest list to `gcr.io/distroless/static:latest`

**Comment by BenTheElder** [comment]
Even with no binary base:
https://github.com/kubernetes/kubernetes/issues/87325

**Comment by dekkagaijin** [comment]
@BenTheElder right, it'd require a manifest list to be pushed to the base image repo. `amd64` is the default, otherwise.

**Comment by spiffxp** [comment]
needs rebase

**Comment by dims** [comment]
Note that before we merge this PR - [Revert "Revert "Revert "[Re-Apply][Distroless] Convert the GCE manifests for master containers."""](https://github.com/kubernetes/kubernetes/pull/83390)  - needs to be Reverted again!

**Comment by dims** [comment]
/close

Thanks a ton @yuwenma we are continuing this work in https://github.com/kubernetes/kubernetes/pull/90674

**Pull request closed without merging**